clippy::suboptimal_flops suggestion can introduce errors with mul_add()

### Summary

The `clippy::suboptimal_flops` lint often suggests to use builtin functions such as `mul_add()` to make code more performant, but this can lead to errors when using the `Complex64` type from the `num` crate together with `f64` values.

### Reproducer

I tried this code:

```rust
use num::complex::Complex64;

fn main() {
    let a = Complex64::new(1.0, 1.0);
    let b = 1.0_f64;
    let c = 1.0;
    
    let d = a + b * c;
    
    println!("{d:?}");
}
```

clippy suggests to rewrite this as:
```rust
use num::complex::Complex64;

fn main() {
    let a = Complex64::new(1.0, 1.0);
    let b = 1.0_f64;
    let c = 1.0;
    
    let d = b.mul_add(c, a);
    
    println!("{d:?}");
}
```

This causes the compiler error:

```
error[E0308]: mismatched types
 --> src/main.rs:8:26
  |
8 |     let d = b.mul_add(c, a);
  |               -------    ^ expected `f64`, found `Complex<f64>`
  |               |
  |               arguments to this method are incorrect
  |
  = note: expected type `f64`
           found struct `Complex<f64>`
note: method defined here
 --> /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/f64.rs:270:12

For more information about this error, try `rustc --explain E0308`.
```

### Version

```text
rustc 1.74.0 (79e9716c9 2023-11-13)
binary: rustc
commit-hash: 79e9716c980570bfd1f666e3b16ac583f0168962
commit-date: 2023-11-13
host: x86_64-unknown-linux-gnu
release: 1.74.0
LLVM version: 17.0.4
```


### Additional Labels

I-suggestion-causes-error

--- a/clippy_lints/src/floating_point_arithmetic.rs
+++ b/clippy_lints/src/floating_point_arithmetic.rs
@@ -496,9 +496,13 @@ fn check_mul_add(cx: &LateContext<'_>, expr: &Expr<'_>) {
             if let BinOpKind::Sub = op { -sugg } else { sugg }
         };
 
-        let (recv, arg1, arg2) = if let Some((inner_lhs, inner_rhs)) = is_float_mul_expr(cx, lhs) {
+        let (recv, arg1, arg2) = if let Some((inner_lhs, inner_rhs)) = is_float_mul_expr(cx, lhs)
+            && cx.typeck_results().expr_ty(rhs).is_floating_point()
+        {
             (inner_lhs, Sugg::hir(cx, inner_rhs, ".."), maybe_neg_sugg(rhs))
-        } else if let Some((inner_lhs, inner_rhs)) = is_float_mul_expr(cx, rhs) {
+        } else if let Some((inner_lhs, inner_rhs)) = is_float_mul_expr(cx, rhs)
+            && cx.typeck_results().expr_ty(lhs).is_floating_point()
+        {
             (inner_lhs, maybe_neg_sugg(inner_rhs), Sugg::hir(cx, lhs, ".."))
         } else {
             return;

--- a/tests/ui/floating_point_mul_add.fixed
+++ b/tests/ui/floating_point_mul_add.fixed
@@ -39,3 +39,21 @@ fn main() {
     // Cases where the lint shouldn't be applied
     let _ = (a * a + b * b).sqrt();
 }
+
+fn _issue11831() {
+    struct NotAFloat;
+
+    impl std::ops::Add<f64> for NotAFloat {
+        type Output = Self;
+
+        fn add(self, _: f64) -> Self {
+            NotAFloat
+        }
+    }
+
+    let a = NotAFloat;
+    let b = 1.0_f64;
+    let c = 1.0;
+
+    let _ = a + b * c;
+}

--- a/tests/ui/floating_point_mul_add.rs
+++ b/tests/ui/floating_point_mul_add.rs
@@ -39,3 +39,21 @@ fn main() {
     // Cases where the lint shouldn't be applied
     let _ = (a * a + b * b).sqrt();
 }
+
+fn _issue11831() {
+    struct NotAFloat;
+
+    impl std::ops::Add<f64> for NotAFloat {
+        type Output = Self;
+
+        fn add(self, _: f64) -> Self {
+            NotAFloat
+        }
+    }
+
+    let a = NotAFloat;
+    let b = 1.0_f64;
+    let c = 1.0;
+
+    let _ = a + b * c;
+}

